File: haskell/acronym/src/Acronym.hs
module Acronym (abbreviate) where

import Data.Char
import Data.List

abbreviate :: String -> String
abbreviate xs
    | elem ':' xs = takeWhile (':' /=) xs
    | otherwise = intercalate "" $ map mapWord $ words xs

mapWord :: String -> String
mapWord word@(y:ys)
    | all isUpper word = [y]
    | elem '-' word = abbreviate $ map (\x -> if x == '-' then ' ' else x) word
    | otherwise = filter isUpper $ toUpper y:ys

File: go/acronym/acronym.go
package acronym

import (
	"strings"
)

// Abbreviate should have a comment documenting it.
func Abbreviate(s string) string {

	words := strings.Fields(strings.Replace(s, "-", " ", -1))

	for i, w := range words {
		words[i] = string(w[0])
	}

	return strings.ToUpper(strings.Join(words, ""))
}
